[Pytorch NNAPI] Add compilation_preference & relax_f32_to_f16 APIs

axitkhurana opened the pull request:
Summary:
compilation_preference is one of:

ANEURALNETWORKS_PREFER_LOW_POWER = 0
ANEURALNETWORKS_PREFER_FAST_SINGLE_ANSWER = 1
ANEURALNETWORKS_PREFER_SUSTAINED_SPEED = 2

relax_f32_to_f16 calls Model_relaxComputationFloat32toFloat16

Test Plan:
Tested on device with nnapi models

* Works with existing exported models
* Works with new exported models with options

Differential Revision: D36433236

dreiss commented on torch/backends/_nnapi/prepare.py:
```diff
@@ -33,14 +39,20 @@ def __init__(
         self.weights = [w.contiguous() for w in self.weights]
         comp = torch.classes._nnapi.Compilation()
-        comp.init(self.ser_model, self.weights)
+        if self.compilation_preference is not None or self.relax_f32_to_f16 is not None:
+            comp.init_with_preferences(self.ser_model, self.weights, self.compilation_preference, self.relax_f32_to_f16)
+        else:
+            comp.init(self.ser_model, self.weights)
+
```
What if exactly one of them is None?  This would fail, right?  I'd just require them to always be present, since it looks like that's what `convert_model_to_nnapi` is doing.

dreiss commented on aten/src/ATen/nnapi/nnapi_bind.h:
```diff
@@ -36,7 +36,19 @@ struct NnapiCompilation : torch::jit::CustomClassHolder {
     ~NnapiCompilation() override = default;
 
-    TORCH_API void init(at::Tensor serialized_model_tensor, std::vector<at::Tensor> parameter_buffers);
+    TORCH_API void init(
```
Maybe comment that this is only necessary to support older models that still call this overload.

dreiss commented on aten/src/ATen/nnapi/nnapi_bind.h:
```diff
@@ -36,7 +36,19 @@ struct NnapiCompilation : torch::jit::CustomClassHolder {
+      std::vector<at::Tensor> parameter_buffers
+    );
+
+    TORCH_API void init_with_preferences(
```
I might just call this "init2" so we can create an "init3" if we need to add more args in the future.  Feels a bit hacky, but I've seen it done successfully before: https://subversion.apache.org/docs/api/latest/group__svn__io__byte__streams.html#ga444db383401ed714a240ff6ca5ca76b3 (svn_stream_copy).

Commit:
[Pytorch NNAPI] Add compilation_preference & relax_f32_to_f16 APIs (#78758)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/78758

compilation_preference is one of:

ANEURALNETWORKS_PREFER_LOW_POWER = 0
ANEURALNETWORKS_PREFER_FAST_SINGLE_ANSWER = 1
ANEURALNETWORKS_PREFER_SUSTAINED_SPEED = 2

relax_f32_to_f16 calls Model_relaxComputationFloat32toFloat16

Test Plan:
Tested on device with nnapi models

* Works with existing exported models
* Works with new exported models with options (1 option, both new options, no options)

Reviewed By: dreiss

Differential Revision: D36433236

fbshipit-source-id: eab74cfcd4186917526d53e321f8a292128c0af8

axitkhurana commented:
@pytorchbot merge -g
